### PR TITLE
feat(sdk): add plan interval cycle toggle

### PR DESCRIPTION
**Pull request by rsbh**
This PR will add support for plan interval cycle toggle.
users can switch plan interval from billing screen.
Screenshot 2024-04-10 at 12 36 22 PM
Screenshot 2024-04-10 at 12 35 19 PM

### CODE REVIEW

File: sdks/js/packages/core/react/components/organization/billing/cycle-switch/index.tsx [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+  const {
+    checkoutPlan,
+    isLoading: isPlanActionLoading,
+    changePlan,
+    verifyPlanChange
+  } = usePlans();
+
+  const nextPlanPrice = nextPlan ? getPlanPrice(nextPlan) : { amount: 0 };
+  const isPaymentRequired =
```
Comment (anujk14) [inline sdks/js/packages/core/react/components/organization/billing/cycle-switch/index.tsx:39]: Can we rename this to `isPaymentMethodRequired` ?

File: sdks/js/packages/core/react/components/organization/billing/cycle-switch/index.tsx [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    _.isEmpty(paymentMethod) && nextPlanPrice.amount > 0;
+
+  const nextPlanIntervalName = getPlanIntervalName(nextPlan);
+
+  const nextPlanMetadata = nextPlan?.metadata as Record<string, number>;
+  const activePlanMetadata = activePlan?.metadata as Record<string, number>;
+
+  const isUpgrade =
```
Comment (anujk14) [inline sdks/js/packages/core/react/components/organization/billing/cycle-switch/index.tsx:47]: TODO for later: Let's move this to a common place?

```
func isUpgrade(plan1, plan2)
Return type: boolean
```

Our logic for determining whether a plan is an upgrade should be the same everywhere. A common method would help us enforce that.
Same for `isDowngrade`, `isPaymentMethodRequired` etc.

File: sdks/js/packages/core/react/components/organization/billing/cycle-switch/index.tsx [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+      getNextPlan(planId);
+    }
+  }, [client, planId]);
+
+  const isLoading = isPlanLoading;
+
+  async function onConfirm() {
+    setCycleSwitching(true);
+    try {
+      if (nextPlan?.id) {
```
Comment (anujk14) [inline sdks/js/packages/core/react/components/organization/billing/cycle-switch/index.tsx:78]: In what scenario will we not have `nextPlan.id`? This method is on clicking "confirm" of the dialogue box, right?

File: sdks/js/packages/core/react/components/organization/billing/upcoming-billing-cycle.tsx [hunk 4]
```diff
@@ -29,6 +34,45 @@ function LabeledBillingData({
...
+        className={billingStyles.linkBtn}
+        onClick={onClick}
+      >
+        Switch to {intervalName}
+      </Button>
+    </div>
+  );
+}
+
+function getSwitchablePlan(plans: V1Beta1Plan[], currentPlan: V1Beta1Plan) {
```
Comment (anujk14) [inline sdks/js/packages/core/react/components/organization/billing/upcoming-billing-cycle.tsx:63]: In the original requirements, are we supposed to show the "Switch to ..." button only for upgrades, or for any cycle switch?

Comment (rsbh) [reply sdks/js/packages/core/react/components/organization/billing/upcoming-billing-cycle.tsx:63]: i think we are showing switch for all cases. (upgrade and downgrade)

File: sdks/js/packages/core/react/components/organization/billing/cycle-switch/index.tsx (hunk 3 above)
Comment (rsbh) [reply sdks/js/packages/core/react/components/organization/billing/cycle-switch/index.tsx:78]: The schema type generated by (swagger-typescript-api)[https://github.com/acacode/swagger-typescript-api] has all keys as optional.
I will fix that in separate PR

File: sdks/js/packages/core/react/components/organization/billing/cycle-switch/index.tsx (hunk 2 above)
Comment (rsbh) [reply sdks/js/packages/core/react/components/organization/billing/cycle-switch/index.tsx:47]: yes, i tried to refactor it, but current PR got big due to that i removed the refactor changes.
right now the same logic is used at 2 places. i will refactor this in future

Comment (anujk14) [reply sdks/js/packages/core/react/components/organization/billing/cycle-switch/index.tsx:47]: Sounds good

File: sdks/js/packages/core/react/components/organization/billing/cycle-switch/index.tsx (hunk 3 above)
Comment (anujk14) [reply sdks/js/packages/core/react/components/organization/billing/cycle-switch/index.tsx:78]: Okay

File: sdks/js/packages/core/react/components/organization/billing/upcoming-billing-cycle.tsx (hunk 4 above)
Comment (anujk14) [reply sdks/js/packages/core/react/components/organization/billing/upcoming-billing-cycle.tsx:63]: Can you confirm on pe-design once? From the design, I also feel we need to show it for both. But based on recent experience, I think we should confirm it once. There is scope for misunderstanding there.

**Review by anujk14 — APPROVED** [review]
lgtm

**Pull request merged**